tensorwaves/physics/helicityformalism/amplitude.py

```diff
@@ -170,238 +223,292 @@ def _initialize_parameters(self, recipe: dict) -> None:
-                dec_prod_fs_ids[1]
-            )
-        else:
-            raise Exception("Helicity Decay expects a dictionary recipe!")
+DynamicsProperties = namedtuple(
+    "DynamicsProperties",
+    [
+        "orbit_angular_momentum",
+        "resonance_mass",
+        "resonance_width",
+        "inv_mass_name",
+        "inv_mass_name_prod1",
+        "inv_mass_name_prod2",
+    ],
+)
 
-    def __call__(self, dataset):
-        return self._call_wrapper(dataset)
 
-    def _with_dynamics(self, dataset):
-        return wigner_capital_d(
-            dataset[self._phi_name],
-            dataset[self._theta_name],
-            0.0,
-            2 * self._j,
-            2 * self._m,
-            2 * self._mprime,
-        ) * self._dynamics_function(
+class _RelativisticBreitWigner:
+    def __init__(
+        self, mass: tf.Variable, width: tf.Variable, inv_mass_name: str
+    ) -> None:
+        self._resonance_mass = mass
+        self._resonance_width = width
+        self._inv_mass_name = inv_mass_name
+
+    def __call__(self, dataset: dict) -> tf.Tensor:
+        return relativistic_breit_wigner(
             dataset[self._inv_mass_name],
             self._resonance_mass,
             self._resonance_width,
         )
 
-    def _without_dynamics(self, dataset):
+
+def _create_relativistic_breit_wigner(
+    dynamics_props: DynamicsProperties,
+) -> Callable:
+    return _RelativisticBreitWigner(
+        dynamics_props.resonance_mass,
+        dynamics_props.resonance_width,
+        dynamics_props.inv_mass_name,
+    )
+
+
+class _NonDynamic:
+    def __call__(self, dataset: dict) -> tf.Tensor:
+        return tf.complex(
+            tf.constant(1.0, dtype=tf.float64),
+            tf.constant(0.0, dtype=tf.float64),
+        )
+
+
+def _create_no_dynamics(_: DynamicsProperties) -> Callable:
+    return _NonDynamic()
+
+
+class _HelicityDecay:
+    def __init__(
+        self, angular_params: "_AngularProperties", dynamics_function: Callable
+    ) -> None:
+        self._params = angular_params
+        self._dynamics_function = dynamics_function
+
+    def __call__(self, dataset: dict) -> tf.Tensor:
         return wigner_capital_d(
-            dataset[self._phi_name],
-            dataset[self._theta_name],
+            dataset[self._params.phi_name],
+            dataset[self._params.theta_name],
             0.0,
-            2 * self._j,
-            2 * self._m,
-            2 * self._mprime,
-        )
+            2 * self._params.j,
+            2 * self._params.m,
+            2 * self._params.mprime,
+        ) * self._dynamics_function(dataset)
+
+
+def _create_helicity_decay(
+    builder: IntensityBuilder, recipe: dict, **kwargs: dict
+) -> Callable:
+    if not isinstance(recipe, dict):
+        raise Exception("Helicity Decay expects a dictionary recipe!")
+    decaying_state = recipe["DecayParticle"]
+    decay_products = recipe["DecayProducts"]["Particle"]
```

Comment:
I think we can apply a small change to the recipe here:
```yaml
DecayParticle:
  Name: J/psi
  Helicity: -1
DecayProducts:
  Particle:
    - Name: f2(1950)
      FinalState: [3, 4]
      Helicity: 0.0
    - Name: gamma
      FinalState: 2
      Helicity: -1
```
can just become
```yaml
DecayParticle:
  Name: J/psi
  Helicity: -1
DecayProducts:
  - Name: f2(1950)
    FinalState: [3, 4]
    Helicity: 0.0
  - Name: gamma
    FinalState: 2
    Helicity: -1
```
and we can write
```python
recipe["DecayProducts"]
```

Reply:
That sounds good. I will open an issue for that.